lib/Conversion/TritonGPUToLLVM/DotOpToLLVM/FMA.cpp

```diff
@@ -1,28 +1,42 @@
 #include "mlir/Support/LLVM.h"
 #include "triton/Conversion/TritonGPUToLLVM/Utility.h"
+#include "triton/Dialect/TritonGPU/IR/Dialect.h"
+#include "triton/Dialect/TritonGPU/Transforms/Utility.h"
 
 using namespace mlir;
 using namespace mlir::triton;
+using namespace ::mlir::triton::gpu;
 
-using ::mlir::triton::gpu::DotOperandEncodingAttr;
+using ::mlir::LLVM::linearize;
+using ::mlir::triton::gpu::expandMatrixOrderWithBatch;
+using ::mlir::triton::gpu::expandMatrixShapeWithBatch;
 using ::mlir::triton::gpu::getShapePerCTA;
-using ::mlir::triton::gpu::NvidiaMmaEncodingAttr;
+using ::mlir::triton::gpu::getSizePerThread;
 
-using ValueTableFMA = std::map<std::pair<int, int>, Value>;
+using ValueTableFMA = std::map<std::tuple<int, int, int>, Value>;
 
 static ValueTableFMA
-getValueTableFromStructFMA(Value val, int K, int n0, int shapePerCTATile,
-                           int sizePerThread,
+getValueTableFromStructFMA(Value val, ArrayRef<unsigned> perTileShape,
+                           unsigned kDim, unsigned nonKDim,
                            ConversionPatternRewriter &rewriter, Location loc,
-                           const LLVMTypeConverter *typeConverter, Type type) {
+                           ArrayRef<unsigned> order) {
   ValueTableFMA res;
   auto elems = unpackLLElements(loc, val, rewriter);
-  int index = 0;
-  for (unsigned k = 0; k < K; ++k) {
-    for (unsigned m = 0; m < n0; m += shapePerCTATile)
-      for (unsigned mm = 0; mm < sizePerThread; ++mm) {
-        res[{m + mm, k}] = elems[index++];
-      }
+  assert(perTileShape.size() == 3);
+  assert(elems.size() == product(perTileShape));
+  assert(kDim == 1 || kDim == 2);
+  assert(nonKDim == 1 || nonKDim == 2);
+  const unsigned bDim = 0;
+
+  for (unsigned idx = 0; idx < elems.size(); ++idx) {
+    unsigned spatialIdx[3];
+    unsigned curIdx = idx;
+    for (auto dim : order) {
+      spatialIdx[dim] = curIdx % perTileShape[dim];
+      curIdx /= perTileShape[dim];
+    }
```

Comment:
Is this logic the same as delinearize()?

Reply:
Yes, there were no function with integer arguments, but I will add it